src/cgroups/memory.rs

```diff
@@ -0,0 +1,336 @@
+                Self::set_i64(kmem, &cgroup_root.join(CGROUP_KERNEL_MEMORY_LIMIT))?;
+            }
+            if let Some(tcp_mem) = memory.kernel_tcp {
+                Self::set_i64(tcp_mem, &cgroup_root.join(CGROUP_KERNEL_TCP_MEMORY_LIMIT))?;
+            }
+
+            OpenOptions::new()
+                .create(false)
+                .write(true)
+                .truncate(false)
+                .open(cgroup_root.join("cgroup.procs"))?
+                .write_all(pid.to_string().as_bytes())?;
+        }
+        Ok(())
+    }
+}
+
+impl Memory {
+    fn get_memory_usage(cgroup_root: &Path) -> Result<u64> {
+        let path = cgroup_root.join(CGROUP_MEMORY_USAGE);
+        let mut contents = String::new();
+        OpenOptions::new()
+            .create(false)
+            .read(true)
+            .open(path)?
+            .read_to_string(&mut contents)?;
+
+        contents = contents.trim().to_string();
+
+        if contents == "max" {
+            return Ok(u64::MAX);
+        }
+
+        let val = contents.parse::<u64>()?;
+        Ok(val)
+    }
+
+    fn get_memory_max_usage(cgroup_root: &Path) -> Result<u64> {
+        let path = cgroup_root.join(CGROUP_MEMORY_MAX_USAGE);
+        let mut contents = String::new();
+        OpenOptions::new()
+            .create(false)
+            .read(true)
+            .open(path)?
+            .read_to_string(&mut contents)?;
+
+        contents = contents.trim().to_string();
+
+        if contents == "max" {
+            return Ok(u64::MAX);
+        }
+
+        let val = contents.parse::<u64>()?;
+        Ok(val)
+    }
+
+    fn get_memory_limit(cgroup_root: &Path) -> Result<i64> {
+        let path = cgroup_root.join(CGROUP_MEMORY_LIMIT);
+        let mut contents = String::new();
+        OpenOptions::new()
+            .create(false)
+            .read(true)
+            .open(path)?
+            .read_to_string(&mut contents)?;
+
+        contents = contents.trim().to_string();
+
+        if contents == "max" {
+            return Ok(i64::MAX);
+        }
+
+        let val = contents.parse::<i64>()?;
+        Ok(val)
+    }
+
+    fn set_i64(val: i64, path: &Path) -> std::io::Result<()> {
+        OpenOptions::new()
+            .create(false)
+            .write(true)
+            .truncate(true)
+            .open(path)?
+            .write_all(val.to_string().as_bytes())?;
+        Ok(())
+    }
+
+    fn set_u64(val: u64, path: &Path) -> std::io::Result<()> {
+        OpenOptions::new()
+            .create(false)
+            .write(true)
+            .truncate(true)
+            .open(path)?
+            .write_all(val.to_string().as_bytes())?;
+        Ok(())
+    }
+
+    fn set_memory(val: i64, cgroup_root: &Path) -> Result<()> {
+        let path = cgroup_root.join(CGROUP_MEMORY_LIMIT);
+
+        match Self::set_i64(val, &path) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                // we need to look into the raw OS error for an EBUSY status
```

Comment:
This was a tough one. Thank you very much 👍

Reply:
@utam0k the challenge was fun! I tried to follow the runc code as cloesly as I could on this one since there were a few caveats, but it was a good learning experience.

Reply:
That was so good! I'm learning from your PR review now too.